### ISSUE TITLE
Better Exceptions

### ISSUE THREAD

**Issue by ribrdb** [Hardmath123/nearley#146]
Is there some way to get better error messages. Like what tokens were expected instead of just which token causes the parse to fail.

**Comment by bobbybee** [comment]
I imagine that for anything reasonably complex, error messages shouldn't
be handled in the grammar -- if the message you're looking for is along
the lines of "Unexpected declaration after number", I imagine you can
find that yourself from the tokenized input.

**Comment by ribrdb** [comment]
I don't understand. Isn't the grammar what defines which combinations of tokens are allowed? How is anything other than the grammar supposed to know what's wrong with a sequence of tokens?

**Comment by Hardmath123** [comment]
@ribrdb, you're completely right. This would be a great feature.

For now, you can already do this "on your own" by inspecting the parse table. If `p` is a parser, then

```
p.table[p.table.length-2].map(function(s) {return s.rule.symbols[s.expect]; }).filter(function(t) {return t.literal || t.test; });
```

gives you the list of tokens that the parser is expecting.

But, I would appreciate a PR or something that integrates this functionality into nearley in a more user-friendly way. Or enough spare time to implement it myself!

**Comment by ribrdb** [comment]
Hmm. So why does nearly sometimes throw an exception and sometimes just return 0 results?
Should this same thing work when there were no results?
I tried this with something like this (which return 0 results with no error)
a = sin(

The expected symbol seems to be ( though, which it should have already matched.

**Comment by Hardmath123** [comment]
"Results" are finished parses at the current location in the input stream. When parsing "1+1", you start with no results, then you get "[ 1 ]", then no results again because "1+" is not well-formed, and then finally "[ (+ 1 1) ]". Zero results does not necessarily mean a parse error, because future input may create results.

Exceptions are thrown when the parser encounters an unexpected token, because this implies that *no* possible parsings will ever exist.

**Comment by nanalan** [comment]
> because future input may create results.

So it's essentially "unexpected EOF," if you've fed all data you want to at that point?

**Comment by Hardmath123** [comment]
Yes.

**Comment by laszlokorte** [comment]
Is there a way to tell nearley that there will be not more input?
Currently I am doing
```
const p = new nearley.Parser(grammar.ParserRules, grammar.ParserStart)
p.feed(el.target.value) // the textarea's value
const result = p.finish()
```

I would be great if result would be either the resulting parse tree or an exception is thrown with the following information: 
1. unexpected token: XXX
2. start position of token
3. end position of token

**Comment by wizzard0** [comment]
@laszlokorte you can do that with testable tokens and feeding list of characters + "EOF"

@Hardmath123 could there be an option that helps default literals to work in this case?

**Comment by tjvr** [comment]
Like most parsing frameworks, simply providing the list of tokens the parser accepts isn't usually very useful. It's certainly not something Nearley should do by default.

Since this issue was opened, our error messages (both with and without a [tokenizer](https://nearley.js.org/docs/tokenizers)) have become much prettier--they now include line & column information, as well as a preview of the line.

```
invalid syntax at line 2 col 3:

  12!
    ^
```
In summary: I'm closing this. :-)
